nevergrad/optimization/optimizerlib.py

```diff
@@ -1203,68 +1203,74 @@ def _internal_tell_candidate(self, candidate: base.Candidate, value: float) -> N
-        list_of_budgets,
-    ) -> None:
-        # Either we have the budget for each algorithm, or the last algorithm uses the rest of the budget, so:
-        assert len(list_of_algorithms) == len(list_of_budgets) or len(list_of_algorithms) == len(list_of_budgets) + 1 
+    Each algorithm is fed with what happened before it.
 
-        # Let us store the two key parameters:
-        self._list_of_budgets = list_of_budgets
-        self._list_of_algorithms = list_of_algorithms
+    Parameters
+    ----------
+    optimizers: list of Optimizer classes
+        the sequence of optimizers to use
+    budgets: list of int
+        the corresponding budgets for each optimizer but the last one
 
+    """
+    _optimizer_class = _Chain
+
+    def __init__(self, optimizers: Sequence[Union[base.OptimizerFamily, Type[base.Optimizer]]],
+                 budgets: Sequence[Union[str, int]]) -> None:
+        # Either we have the budget for each algorithm, or the last algorithm uses the rest of the budget, so:
+        self.budgets = tuple(budgets)
+        self.optimizers = tuple(optimizers)
+        assert len(self.optimizers) == len(self.budgets) + 1
+        assert all(x in ("dimension", "num_workers") or x > 0 for x in self.budgets)  # type: ignore
+        super().__init__()
 
-    # From the user point of view, the call below is the creation of the optimizer.
-    def __call__(self, instrumentation: Union[int, Instrumentation], budget: Optional[int] = None, num_workers: int = 1) -> base.Optimizer:
 
-        list_of_algorithms = self._list_of_algorithms
-        list_of_budgets = self._list_of_budgets
-
-        # We create the class we want to use.
-        class chain(base.Optimizer):
-
-            def __init__(self, instrumentation: Union[int, Instrumentation], budget: Optional[int] = None, num_workers: int = 1) -> None:
-                super().__init__(instrumentation, budget=budget, num_workers=num_workers)
-                self._list_of_budgets = list_of_budgets
-                self._list_of_algorithms = [l(instrumentation, budget, num_workers) for l in list_of_algorithms]
-                # -1 in the list of budgets means "num_workers".
-                # -2 in the list of budgets means "dimension".
-                for i in range(len(self._list_of_budgets)):
-                    if self._list_of_budgets[i] == -1:
-                        self._list_of_budgets[i] = num_workers
-                    if self._list_of_budgets[i] == -2:
-                        self._list_of_budgets[i] = self.dimension
-
-            def _internal_ask_candidate(self) -> base.Candidate:
-                # Which algorithm are we playing with ?
-                algo_index = 0
-                for b in self._list_of_budgets:
-                    if self._num_ask >= b:
-                        algo_index += 1
-                # Ok, that algorithm chooses the next point to evaluate.
-                return self._list_of_algorithms[algo_index].ask()
```

Comment:
Here is one of the reasons I preferred avoiding propagating cheap constraints through `ask`: the constraints would have needed to be propagated to subalgorithms as well. With the constraints within the instrumentation, this is automatic

Reply:
Right.